### ISSUE TITLE
RPC use case

### ISSUE THREAD

**Issue by AArnott** [eiriktsarpalis/PolyType#80]
StreamJsonRpc currently is not trim-safe, I suppose, because given a type (which may be a class or interface) it reflects over it to discover methods on it to set up RPC dispatch maps for it. It also generates a proxy for the client side.

If PolyType offered a `[GenerateBehavior]` companion attribute for `[GenerateShape]`, I imagine it could source-generate a reflection-free definition of the _methods_ rather than the data, allowing StreamJsonRpc to consume it as `ITypeBehavior` instead of `ITypeShape`.

What do you think?

**Comment by eiriktsarpalis** [comment]
In the StreamJsonRpc use case, are the signatures of the discovered methods fixed or could they accept parameters of any type? With a hypothetical `[GenerateBehavior]` gesture, would this also need to transitively generate the shapes of all the parameter types as well?

If not, would a linker annotation (such as `DynamicallyAccessedMembers`) suffice?

**Comment by AArnott** [comment]
The methods may use whatever parameters they like. Generic _methods_ are not supported.
StreamJsonRpc would then be on the hook to serialize each argument (using Nerdbank.MessagePack, presumably) so yes, we'd want a type shape provider to know about each parameter type. What's more, we wouldn't want each parameter type to have to be decorated or require an explicit witness class. It should "just work" by virtue of being referenced from the `[GenerateBehavior]` decorated type.

Just to be thorough, not every parameter or return type are directly serializable either. `CancellationToken` gets special, non-serialized treatment, as do other types like `Stream`, `PipeReader`/`PipeWriter`, `IProgress<T>` and `Task` return types.

```cs
[GenerateBehavior]
interface IRpcContract
{
    Task DoSomethingAsync(int count, string foo, MyCustomType value, CancellationToken cancellationToken);
    Task DoItSlowly(IProgress<MyUpdates> progress);
}
```

We'd probably need to assume that the `IBehaviorShape<IRpcContract>` object had an `ITypeShapeProvider` property on it that we could use to get all the type shapes.

Using `[DynamicallyAccessedMembers]` would presumably work to make it trim-safe. But the means to get all the `ITypeShape` objects for the parameter and return types can't possibly be provable by generic type constraints the way it is today with `IShapeable<T>`. Even if we just provide an `ITypeShapeProvider` that is expected to know _all_ the necessary shapes for the parameter/return types, we'd want an experience that wasn't as tedious for the user as to create a witness type for every single parameter type that may appear on the interface just to populate the shape provider. So something in the system that can trace those type references and auto-select those types for shape generation would be _very_ useful.

**Comment by AArnott** [comment]
I'm drafting out what this might look like here: 
https://github.com/AArnott/PolyBehavior/tree/master/PolyBehavior

I'm mostly modeling it after PolyType's existing patterns. But among the experiments, I'm mixing in more public concrete types rather than keeping to interfaces.

In the case of RPC, a given method parameter may be a _shape_ or a _behavior_ (a serialized value or a marshaled interface). I can't tell until I look at the parameter type. But this suggests that shapes and behaviors are somewhat intertwined.

**Comment by eiriktsarpalis** [comment]
Here's a rough sketch of a method shape API:
```C#
public interface IMethodShape
{
    Type DeclaringType { get; }
    bool IsStatic { get; }
    ICustomAttributeProvider? AttributeProvider { get; }
    IReadOnlyList<IParameterShape> Parameters { get; }
    object? Accept(MethodShapeVisitor visitor, object? state = null);
}

public interface IMethodShape<TArguments, TReturnType>
{
    Func<object?, TArguments, ValueTask<TReturnType>> GetInvoker();
}

public abstract class MethodShapeVisitor : TypeShapeVisitor
{
    public abstract object? VisitMethod<TArguments, TReturnType>(IMethodShape<TArguments, TReturnType> method, object? state = null);
}
```
with the following breaking change to `ITypeShapeProvider`:
```diff
-public interface ITypeShapeProvider
-{
-    ITypeShape? GetShape(Type type);
- }
+public interface IShapeProvider
+{
+    ITypeShape? GetTypeShape(Type type);
+    IMethodShape? GetMethodShape(MethodInfo methodInfo);
+}
```
The source generator can generate shapes for methods using a `[GenerateShape]` annotation on individual method declarations.

**Comment by AArnott** [comment]
> The source generator can generate shapes for methods using a `[GenerateShape]` annotation on individual method declarations.

What about whole types? For an RPC interface, we want shapes for all methods.
But if `[GenerateShape]` applied to a class produced a data shape and a behavior shape, that's a lot of code gen when maybe all you wanted was serializability. It seems their ought to be _some_ way to indicate or infer whether you need behavior and/or data shapes.

> public interface IMethodShape

Will `IParameterShape<T>` include an `ITypeShape<T>`  member?
Maybe `IMethodShape` should include an `ITypeShape ReturnType` member, and `IMethodShape<TArg, TRet>` should have `new `ITypeShape<TRet> ReturnType`.

> `IMethodShape<TArguments, TReturnType>`

I wonder if it should have a `TDeclaringType` on it as well so that the `GetInvoker()` method can return a proper delegate equivalent of `Func<ref TDeclaringType?, TArguments, ValueTask<TReturnType>>`, so that methods on structs can be invoked without boxing, and allowing the struct method to modify the struct (since boxing would effectively render it useless to mutate).
I wonder if `TArguments` should be passed by `ref` as well, to support `ref` parameters or if we don't support `ref` parameters, maybe use the `in` modifier at least to avoid copying memory for all those arguments again.

**Comment by eiriktsarpalis** [comment]
> > public interface IMethodShape
> 
> Will `IParameterShape<T>` include an `ITypeShape<T>` member?

Yes. My expectation is that this will be the new name for `IConstructorParameterShape<T>` repurposed to also work for methods.

> I wonder if it should have a TDeclaringType on it as well so that the GetInvoker() method can return a proper delegate equivalent of Func<ref TDeclaringType?, TArguments, ValueTask<TReturnType>>, so that methods on structs can be invoked without boxing, and allowing the struct method to modify the struct (since boxing would effectively render it useless to mutate).

Possibly yeah.

> I wonder if TArguments should be passed by ref as well, to support ref parameters or if we don't support ref parameters, maybe use the in modifier at least to avoid copying memory for all those arguments again.

Yes, should be. Basically it would be extended the constructor model.

**Comment by eiriktsarpalis** [comment]
Perhaps the answer though is exposing a more lightweight set of APIs. The [`AIFunctionFactory`](https://github.com/dotnet/extensions/blob/main/src/Libraries/Microsoft.Extensions.AI/Functions/AIFunctionFactory.cs) implementation in Microsoft.Extensions.AI has shown that it's possible to implement AOT safe RPC provided only that the contracts of the parameter types are known (albeit with a bit of boxing since the arguments as passed as `object?`). So a lightweight solution might simply involve having `[GenerateShape]` on a method include all parameter types in the current source generation.

**Comment by eiriktsarpalis** [comment]
I expect that would also include `Dispose()` provided the interface is `IDisposable`. Not saying this is necessarily a problem, just pointing out that it could generate unexpected things.

**Comment by AArnott** [comment]
And I need to see events too. In particular, I need to be able to add event handlers to those events. For what StreamJsonRpc requires, we support `EventHandler` and `EventHandler<T>` event types, where `T` has _no_ generic constraints applied.
Today, we do this by declaring a generic method that will serve as the handler for `EventHandler<T>`, and we call `MakeGenericMethod` at runtime. That constitutes dynamic code however, and we need a substitute approach.

**Comment by AArnott** [comment]
I think the associated types feature we already have will take care of the generic events requirement, provided you create a shape for EventArgs<T>.

**Comment by AArnott** [comment]
> Including explicitly implemented interface methods?

No, I'm not even talking about implementations. I'm just talking about the interfaces:

```cs
interface IFoo
{
    void Foo();
}

[GenerateBehavior]
interface IBar : IFoo
{
   void Bar();
}
```

When you ask for methods on `IBar`, it will _only_ produce `Bar()`. Yet if `IBar` has a `[GenerateBehavior]` attribute on it, I fully expect _all_ methods on that interface and other interfaces it derives from to be included. So I expect `Bar()`, `Foo()` method shapes to be enumerated in the 'behavior' data structure.

In the end, this behavior structure will be used to wire up an RPC target. So consider this class:

```cs
class Server : IBar
{
   public void Foo() { /*... */ }
   public void Bar() { /*... */ }
}
```

If the RPC target were added like this:

```cs
rpc.AddTarget<IBar>(new Server());
```

Then we'll use the behavior of `IBar` no matter what methods are on `Server` or whether they are explicitly implemented or not. So in the above example, we'd get 2 methods wired up.

Now let's take a different case: 

```cs
[GenerateBehavior]
class Server : IBar
{
   public void Foo() { /*... */ }
   void IBar.Bar() { /*... */ }
}
```

`Bar` here is an explicit method implementation. Adding `Server` as an RPC target:

```cs
rpc.AddTarget<IBar>(new Server()); // two methods wired up
rpc.AddTarget<Server>(new Server()); // one method wired up
```

In the first case, we still get two methods because `IBar` (an interface) is the behavior target, and interface members are all 'public'.
In the second case, we get only one method because `Server` (a _class_) being the behavior target, only public methods are considered.

**Comment by eiriktsarpalis** [comment]
That matches my expectation, and is also how properties are being resolved in similar circumstances.
